### ISSUE TITLE
add parameter for redirecting to SSL

### ISSUE THREAD

**Issue by cusco** [mattiasgeniar/varnish-4.0-configuration-templates#36]
Hi,

I've set up this vcl template. Thank you very much, its very useful.

Now, my setup is (client) --https--> apache 443 --proxy pass --> varnish 80 --> apache 8080

so apache does the ssl termination and proxies into varnish.
varnish then requests apache non https port

this works and requires in apache non https vhost:
SetEnvIf X_FORWARDED_PROTO "^https$" HTTPS=on

on vhost https I have:
        ProxyPreserveHost On
        ProxyPass / http://127.0.0.1:80/
#       ProxyPassReverse / http://localhost:80/
        RequestHeader set X-Forwarded-Port "443"
        RequestHeader set X-Forwarded-Proto "https"

and for wordpress users, in wp-config.php i have:
define('FORCE_SSL_ADMIN', true);
// in some setups HTTP_X_FORWARDED_PROTO might contain
// a comma-separated list e.g. http,https
// so check for https existence
if (strpos($_SERVER['HTTP_X_FORWARDED_PROTO'], 'https') !== false)
       $_SERVER['HTTPS']='on';

before loading wp-settings.php

Now, varnish should have a way of redirecting to https if client is not using it.. something like:

if(client.ip != "127.0.0.1"){
  set req.http.x-redir = "https://" + req.http.host + req.url;
  return(synth(850, "Moved permanently"));
}

and 
sub vcl_synth {
        if (resp.status == 850) {
                set resp.http.Location = req.http.x-redir;
                set resp.status = 302;
                return (deliver);
        }
}

I took this code from somewhere on the internet, but i'm afraid its not quite right..

Would it be possible for you to implement such feature?

**Comment by cusco** [comment]
Also, I'm guessing there would be a list of hosts, and then match req.http.host with that lookup...

**Comment by mattiasgeniar** [comment]
Hi!

This shouldn't be too hard to implement in Varnish, it's essentially the same logical condition as your Apache vhost redirect on port :8080. Something like this should work, in vcl_recv.

```
  if (req.http.X-Forwarded-Proto !~ "(?i)https") {
    # This is not an HTTPs connection via Apache
    if (req.http.host ~ "^(domain1.tld|domain2.tld|domain3.tld)$ ") {
      return(synth(850, "https://" + req.http.host + req.url));
    }
  }
```

Could you give that a try?

**Comment by mattiasgeniar** [comment]
> Should I set something regarding the x-forwarded-proto in varnish at anytime?

No, if the Apache proxy sets it, that check in Varnish will work just fine.

As for the curl tip: I should definitely do that!
